### PR TITLE
Provide a more aggressive Roaring64Map::fastunion() implementation

### PR DESCRIPTION
**Pull request by kosak**
`Roaring64Map::operator|=` has this comment
```
* See also the fastunion function to aggregate many bitmaps more quickly.
```

Yet, when we go to the `fastunion()` function we find this comment in the implementation:
```
// not particularly fast
```

Here I propose a priority queue-based implementation which was probably more in line with what was intended.

### CODE REVIEW

**Comment by SLieve** [comment]
I think we should benchmark this, both so we can see how much this provides an improvement over the previous implementation but also for any future optimizations.

**Comment by kosak** [comment]
For this benchmark at least, the results look promising (~2.5 times faster)

```
19:45 $ ./benchmarks/fastunion_benchmark 
*** Legacy fastunion ***
Iteration 0: 2257664 per map
Iteration 1: 2273201 per map
Iteration 2: 2279489 per map
Iteration 3: 2309659 per map
Iteration 4: 2254893 per map
Iteration 5: 2183422 per map
Iteration 6: 2156359 per map
Iteration 7: 2243589 per map
Iteration 8: 2264228 per map
Iteration 9: 2234766 per map
*** New fastunion() ***
Iteration 0: 821548 per map
Iteration 1: 833374 per map
Iteration 2: 855961 per map
Iteration 3: 841438 per map
Iteration 4: 830876 per map
Iteration 5: 894003 per map
Iteration 6: 871780 per map
Iteration 7: 915697 per map
Iteration 8: 861747 per map
Iteration 9: 831667 per map
```

File: cpp/roaring64map.hh [hunk 1]
```diff
@@ -1237,12 +1238,120 @@ public:
...
+                // {candidate_current_iter, candidate_end_iter} into the PQ
+                // (unless it has reached its end).
+                ++candidate_current_iter;
+                if (candidate_current_iter != candidate_end_iter) {
+                    pq.emplace(candidate_current_iter, candidate_end_iter);
+                }
+            }
+
+            // Use the fast inner union to combine these.
+            auto innerResult = roaring_bitmap_or_many(group_bitmaps.size(),
```
Comment (SLieve) [inline cpp/roaring64map.hh:1347]: Nit: I'd suggest using `auto *inner_result` to indicate a pointer.

**Comment by lemire** [comment]
Please check the ubuntu-ci error:

```
236
[ 36%] Building CXX object benchmarks/CMakeFiles/fastunion_benchmark.dir/fastunion_benchmark.cpp.o
[237](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:238)
/home/runner/work/CRoaring/CRoaring/benchmarks/fastunion_benchmark.cpp:1: warning: "_GNU_SOURCE" redefined
[238](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:239)
    1 | #define _GNU_SOURCE
[239](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:240)
      | 
[240](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:241)
<command-line>: note: this is the location of the previous definition
[241](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:242)
In file included from /home/runner/work/CRoaring/CRoaring/cpp/roaring64map.hh:23,
[242](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:243)
                 from /home/runner/work/CRoaring/CRoaring/benchmarks/fastunion_benchmark.cpp:5:
[243](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:244)
/home/runner/work/CRoaring/CRoaring/cpp/roaring.hh:17:2: error: #error "__cpp_exceptions should be defined"
[244](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:245)
   17 | #error "__cpp_exceptions should be defined"
[245](https://github.com/RoaringBitmap/CRoaring/actions/runs/3457725660/jobs/5771433875#step:3:246)
      |  ^~~~~
```

**Comment by lemire** [comment]
(Great idea.)

File: cpp/roaring64map.hh [hunk 2]
```diff
@@ -1237,12 +1238,120 @@ public:
...
+        // current_iter on all the affected entries and then repeats.
+
+        // There is an entry in our priority queue for each of the 'n' inputs.
+        // For a given Roaring64Map, we look at its underlying 'roarings'
+        // std::map, and take its begin() and end(). This forms our half-open
+        // interval [current_iter, end_iter), which we keep in the priority
+        // queue as a std::pair. These entries are updated as our algorithm
+        // progresses. When the interval becomes empty, it is removed from the
+        // priority queue.
+        typedef std::pair<roarings_t::const_iterator, roarings_t::const_iterator> pq_entry_t;
```
[the remark this answers is not in the document]
Comment (kosak) [reply cpp/roaring64map.hh]: Great suggestion, thanks. (done)

File: cpp/roaring64map.hh (hunk 1 above)
[the remark this answers is not in the document]
Comment (kosak) [reply cpp/roaring64map.hh]: done

File: cpp/roaring64map.hh [hunk 3]
```diff
@@ -1237,12 +1238,120 @@ public:
...
+            group_bitmaps.clear();
+            while (!pq.empty()) {
+                auto candidate_current_iter = pq.top().first;
+                auto candidate_end_iter = pq.top().second;
+
+                auto candidate_key = candidate_current_iter->first;
+                const auto &candidate_bitmap = candidate_current_iter->second;
+
+                // This element will either be in the group (having
+                // key == top_key) or it will not be in the group (having
```
[the remark this answers is not in the document]
Comment (kosak) [reply cpp/roaring64map.hh]: done

File: cpp/roaring64map.hh [hunk 4]
```diff
@@ -1237,12 +1238,120 @@ public:
...
+        // A reusable vector that holds the pointers to the inner bitmaps that
+        // we pass to the underlying 32-bit fastunion operation.
+        std::vector<const roaring_bitmap_t*> group_bitmaps;
+
+        // Summary of the algorithm:
+        // 1. While the pq is not empty:
+        //    A. Get its lowest key. Call this group_key
+        //    B. While the lowest entry in the map has a key equal to group_key:
+        //       1. Remove this entry (the pair {current_iter, end_iter}) from
+        //          the map.
```
[the remark this answers is not in the document]
Comment (kosak) [reply cpp/roaring64map.hh]: done

**Comment by kosak** [comment]
While I don't exactly know what I'm doing here, I believe the latest commit I just pushed should fix the build breakage on `Ubuntu CI / ubuntu-noexcept-ci.yml`

(the fact that *all* builds are breaking at the moment is addressed by https://github.com/RoaringBitmap/CRoaring/pull/411 )

**Review by SLieve — APPROVED** [review]

**Comment by lemire** [comment]
I'll merge and fix.

**Pull request merged**